src/bartender/web/api/job/views.py

```diff
@@ -274,3 +280,118 @@ async def retrieve_job_directories_from_path(
+    # https://github.com/PyFilesystem/pyfilesystem2/pull/464
+) -> FileResponse:
+    """Download contents of job directory as archive.
+
+    Args:
+        job_dir: The job directory.
+        background_tasks: FastAPI mechanism for post-processing tasks
+        archive_format: Format to use for archive. Supported formats are '.zip', '.tar',
+            '.tar.xz', '.tar.gz', '.tar.bz2'
+        exclude: list of filename patterns that should be excluded from archive.
+        exclude_dirs: list of directory patterns that should be excluded from archive.
+
+    Returns:
+        FileResponse: Archive containing the content of job_dir
+
+    """
+    dst_fs = _parse_archive_format(archive_format)
+    archive_fn = str(job_dir.with_suffix(archive_format))
+    with (  # noqa: WPS316
+        OSFS(str(job_dir)) as src,
+        dst_fs(archive_fn, write=True) as dst,
+    ):
+        await async_wrap(copy_fs)(
+            src,
+            dst,
+            walker=Walker(exclude=exclude, exclude_dirs=exclude_dirs),
+        )
+
+    background_tasks.add_task(_remove_archive, archive_fn)
+
+    return_fn = Path(archive_fn).name
+    return FileResponse(archive_fn, filename=return_fn)
+
+
+def _parse_archive_format(archive_format: str) -> Union[Type[ZipFS], Type[TarFS]]:
+    if archive_format == ".zip":
+        return ZipFS
+    return TarFS
+
+
+@router.get("/{jobid}/archive/{path:path}")
+async def retrieve_job_subdirectory_as_archive(  # noqa: WPS211
+    path: str,
+    job_dir: CurrentCompletedJobDir,
+    background_tasks: BackgroundTasks,
+    archive_format: ArchiveFormats = ".zip",
+    exclude: Optional[list[str]] = Query(default=None),
+    exclude_dirs: Optional[list[str]] = Query(default=None),
+) -> FileResponse:
+    """Download job output as archive.
+
+    Args:
+        path: Sub directory inside job directory to start from.
+        job_dir: The job directory.
+        background_tasks: FastAPI mechanism for post-processing tasks
+        archive_format: Format to use for archive. Supported formats are '.zip',
+            '.tar', '.tar.xz', '.tar.gz', '.tar.bz2'
+        exclude: list of filename patterns that should be excluded from archive.
+        exclude_dirs: list of directory patterns that should be excluded from archive.
+
+    Raises:
+        HTTPException: When path is not found or is outside job directory.
+
+    Returns:
+        FileResponse: Archive containing the output of job_dir
+
+    """
+    # TODO refactor subdir validation function?
+    try:
+        job_output_dir = (job_dir / path).expanduser().resolve(strict=True)
+        if not job_output_dir.is_relative_to(job_dir):
+            raise FileNotFoundError()
+        if not job_output_dir.is_dir():
+            raise FileNotFoundError()
+    except FileNotFoundError as exc:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="File not found",
+        ) from exc
```

Comment:
This looks very similar to retrieve_job_directories_from_path(), maybe we can extract it to function and lessen code duplication.

Reply:
Yes, I thought the same. Just a function or could we make some kind of validator for subdirectories out of it?